equinox/nn/_attention.py

```diff
@@ -308,6 +390,39 @@ def __call__(
                     "process_heads must not change the shape of the heads."
                 )
 
+        if state is sentinel:
+            causal_mask_offset = 0
+        else:
+            key_state, value_state, index = state.get(self.autoregressive_index)
+            key_state = lax.dynamic_update_slice_in_dim(
+                key_state, key_heads, index, axis=0
+            )
+            value_state = lax.dynamic_update_slice_in_dim(
+                value_state, value_heads, index, axis=0
+            )
```

Comment:
By the way, I think there is actually a subtle bug in these two lines: when `index` is close to the end, then it gets pushed back so that all of the inserted array is inserted (rather than simply truncating the inserted array to only insert the bits that fit).

I don't know if that's something that can be resolved efficiently, though.

Reply:
Ah I see. Well, the solution is probably to truncate `key|value_state` to account for that before passing them into the `dynamic_update_slice_in_dim` function. I'll see what I can do here.

Reply:
Alternatively, we might want to employ the FiFo strategy and roll the key and value state arrays to make room for the new entries.

Reply:
I've added rolling; but I'll need to train something to see if it still works fine this way. I'll keep you posted.